### PR TITLE
[core] Return event subscription in the `addListener` function

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

`addListener` should return an event subscription with a `remove` function to make revoking listeners easier.

# How

Return a subscription object from `addListener` that contains a `remove` function

# Test Plan

Added some new native tests on iOS to cover this. Equivalent tests on Android are probably not necessary as these tests are mostly just relying on JS execution / Hermes implementation.

### CODE REVIEW

**Review by lukmccall — APPROVED** [review]

File: packages/expo-modules-core/common/cpp/EventEmitter.cpp [hunk 1]
```diff
@@ -71,6 +71,11 @@ NativeState::Shared NativeState::get(jsi::Runtime &runtime, jsi::Object &object,
   return nullptr;
 }
 
+#pragma mark - SubscriptionNativeState
+
+SubscriptionNativeState::SubscriptionNativeState(jsi::Object &emitter, jsi::Function &listener)
+  : jsi::NativeState(), emitter(std::move(emitter)), listener(std::move(listener)) {}
```
Comment (Kudo) [inline packages/expo-modules-core/common/cpp/EventEmitter.cpp:77]: using move semantics for referenced `jsi::Object &emitter` looks suspicious. is that intentional?

File: packages/expo-modules-core/common/cpp/EventEmitter.cpp [hunk 2]
```diff
@@ -89,6 +94,69 @@ void callObservingFunction(jsi::Runtime &runtime, jsi::Object &object, const cha
     });
 }
 
+void addListener(jsi::Runtime &runtime, jsi::Object &emitter, std::string eventName, const jsi::Function &listener) {
```
Comment (Kudo) [inline packages/expo-modules-core/common/cpp/EventEmitter.cpp:97]:
```suggestion
void addListener(jsi::Runtime &runtime, jsi::Object &emitter, const std::string &eventName, const jsi::Function &listener) {
```

File: packages/expo-modules-core/common/cpp/EventEmitter.cpp [hunk 3]
```diff
@@ -89,6 +94,69 @@ void callObservingFunction(jsi::Runtime &runtime, jsi::Object &object, const cha
...
+  if (NativeState::Shared state = NativeState::get(runtime, emitter, true)) {
+    state->listeners.add(runtime, eventName, listener);
+
+    if (state->listeners.listenersCount(eventName) == 1) {
+      callObservingFunction(runtime, emitter, "startObserving", eventName);
+    }
+  }
+}
+
+void removeListener(jsi::Runtime &runtime, jsi::Object &emitter, std::string eventName, const jsi::Function &listener) {
```
Comment (Kudo) [inline packages/expo-modules-core/common/cpp/EventEmitter.cpp:107]:
```suggestion
void removeListener(jsi::Runtime &runtime, jsi::Object &emitter, const std::string &eventName, const jsi::Function &listener) {
```

File: packages/expo-modules-core/common/cpp/EventEmitter.cpp [hunk 4]
```diff
@@ -89,6 +94,69 @@ void callObservingFunction(jsi::Runtime &runtime, jsi::Object &object, const cha
...
+
+    state->listeners.remove(runtime, eventName, listener);
+
+    if (listenersCountBefore >= 1 && state->listeners.listenersCount(eventName) == 0) {
+      callObservingFunction(runtime, emitter, "stopObserving", eventName);
+    }
+  }
+}
+
+void removeAllListeners(jsi::Runtime &runtime, jsi::Object &emitter, std::string eventName) {
```
Comment (Kudo) [inline packages/expo-modules-core/common/cpp/EventEmitter.cpp:119]:
```suggestion
void removeAllListeners(jsi::Runtime &runtime, jsi::Object &emitter, const std::string &eventName) {
```

File: packages/expo-modules-core/common/cpp/EventEmitter.cpp [hunk 5]
```diff
@@ -89,6 +94,69 @@ void callObservingFunction(jsi::Runtime &runtime, jsi::Object &object, const cha
...
+
+    state->listeners.removeAll(eventName);
+
+    if (listenersCountBefore >= 1) {
+      callObservingFunction(runtime, emitter, "stopObserving", eventName);
+    }
+  }
+}
+
+void emitEvent(jsi::Runtime &runtime, jsi::Object &emitter, std::string eventName, const jsi::Value *args, size_t count) {
```
Comment (Kudo) [inline packages/expo-modules-core/common/cpp/EventEmitter.cpp:131]:
```suggestion
void emitEvent(jsi::Runtime &runtime, jsi::Object &emitter, const std::string &eventName, const jsi::Value *args, size_t count) {
```

File: packages/expo-modules-core/common/cpp/EventEmitter.cpp [hunk 6]
```diff
@@ -89,6 +94,69 @@ void callObservingFunction(jsi::Runtime &runtime, jsi::Object &object, const cha
...
+  if (NativeState::Shared state = NativeState::get(runtime, emitter, false)) {
+    // Pass the arguments without the first that is the event name.
+    const jsi::Value *payloadArgs = count > 1 ? &args[1] : nullptr;
+    state->listeners.call(runtime, eventName, emitter, payloadArgs, count - 1);
+  }
+}
+
+jsi::Object createEventSubscription(jsi::Runtime &runtime, std::string eventName, jsi::Object emitter, jsi::Function listener) {
```
Comment (Kudo) [inline packages/expo-modules-core/common/cpp/EventEmitter.cpp:139]:
```suggestion
jsi::Object createEventSubscription(jsi::Runtime &runtime, const std::string &eventName, jsi::Object emitter, jsi::Function listener) {
```

**Review by Kudo — APPROVED** [review]
nit: passing string references

File: packages/expo-modules-core/common/cpp/EventEmitter.h [hunk 7]
```diff
@@ -16,7 +16,10 @@ namespace expo::EventEmitter {
 class Listeners {
 private:
   friend class NativeState;
-  friend void installClass(jsi::Runtime &runtime);
+  friend void addListener(jsi::Runtime &runtime, jsi::Object &emitter, std::string eventName, const jsi::Function &listener);
+  friend void removeListener(jsi::Runtime &runtime, jsi::Object &emitter, std::string eventName, const jsi::Function &listener);
+  friend void removeAllListeners(jsi::Runtime &runtime, jsi::Object &emitter, std::string eventName);
+  friend void emitEvent(jsi::Runtime &runtime, jsi::Object &emitter, std::string eventName, const jsi::Value *args, size_t count);
```
Comment (tsapeta) [inline packages/expo-modules-core/common/cpp/EventEmitter.h:22]:
```suggestion
  friend void addListener(jsi::Runtime &runtime, jsi::Object &emitter, const std::string &eventName, const jsi::Function &listener);
  friend void removeListener(jsi::Runtime &runtime, jsi::Object &emitter, const std::string &eventName, const jsi::Function &listener);
  friend void removeAllListeners(jsi::Runtime &runtime, jsi::Object &emitter, const std::string &eventName);
  friend void emitEvent(jsi::Runtime &runtime, jsi::Object &emitter, const std::string &eventName, const jsi::Value *args, size_t count);
```

**Pull request merged**